cli/lsp/analysis.rs

```diff
@@ -948,6 +984,136 @@ fn prepend_whitespace(content: String, line_content: Option<String>) -> String {
   }
 }
 
+/// Get LSP range from the provided SWC start and end locations.
+fn get_range_from_loc(start: &Loc, end: &Loc) -> lsp::Range {
+  lsp::Range {
+    start: lsp::Position {
+      line: (start.line - 1) as u32,
+      character: (start.col_display + 1) as u32,
+    },
+    end: lsp::Position {
+      line: (end.line - 1) as u32,
+      character: (end.col_display - 1) as u32,
+    },
+  }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct TextRange {
+  pub range: lsp::Range,
+  pub text: String,
+}
+
+impl TextRange {
+  fn includes(&self, position: &lsp::Position) -> bool {
+    (position.line > self.range.start.line
+      || position.line == self.range.start.line
+        && position.character >= self.range.start.character)
+      && (position.line < self.range.end.line
+        || position.line == self.range.end.line
+          && position.character <= self.range.end.character)
+  }
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct ImportRanges(Vec<TextRange>);
+
+impl ImportRanges {
+  pub fn contains(&self, position: &lsp::Position) -> Option<TextRange> {
+    self.0.iter().find(|r| r.includes(position)).cloned()
+  }
+}
+
+struct ImportRangeCollector {
+  import_ranges: ImportRanges,
+  source_map: Rc<SourceMap>,
+}
+
+impl ImportRangeCollector {
+  pub fn new(source_map: Rc<SourceMap>) -> Self {
+    Self {
+      import_ranges: ImportRanges::default(),
+      source_map,
+    }
+  }
+
+  pub fn take(self) -> ImportRanges {
+    self.import_ranges
+  }
+}
+
+impl Visit for ImportRangeCollector {
+  fn visit_import_decl(
+    &mut self,
+    node: &swc_ast::ImportDecl,
+    _parent: &dyn Node,
+  ) {
+    let start = self.source_map.lookup_char_pos(node.src.span.lo);
+    let end = self.source_map.lookup_char_pos(node.src.span.hi);
+    self.import_ranges.0.push(TextRange {
+      range: get_range_from_loc(&start, &end),
+      text: node.src.value.to_string(),
+    });
+  }
+
+  fn visit_named_export(
+    &mut self,
+    node: &swc_ast::NamedExport,
+    _parent: &dyn Node,
+  ) {
+    if let Some(src) = &node.src {
+      let start = self.source_map.lookup_char_pos(src.span.lo);
+      let end = self.source_map.lookup_char_pos(src.span.hi);
+      self.import_ranges.0.push(TextRange {
+        range: get_range_from_loc(&start, &end),
+        text: src.value.to_string(),
```

Comment:
We may not want to have this struct have a general `TextRange` name because this allows for the value of `text` to be set manually, which could cause the text value to not align with the range. People should probably derive the text from the source file based on the span/range (for example, `src.value` does not always equal the text seen in a string literal that the actual range could be based on because `src.value` could contain decoded escape sequences). We may want to consider renaming `TextRange` to something else if we do this (ex. perhaps just a specific `ImportRange`).

Reply:
Yeah, I tried to make it more "generic" but realise it causes problems, like the mis-alignment between the range and the text.  I have it before as `specifier`.  I should probably restore that.  I don't want to deal just trying to copy out the string from the text at this point, but maybe I can.